File: tools/docs-generator/lib/physics/body/body_prefab_xml_reader.py
import xml.etree.ElementTree as EXml

from lib.physics.body.body_prefab import BodyPrefab
from lib.utility.string_functions import string_to_boolean


class BodyPrefabXmlReader(object):
    BODY_ELEMENT = "body"

    BODY_TYPE_ELEMENT = "type"
    ACTIVE_ELEMENT = "active"
    AWAKE_ELEMENT = "awake"
    ALLOW_SLEEP_ELEMENT = "allowSleep"
    FIXED_ROTATION_ELEMENT = "fixedRotation"
    BULLET_ELEMENT = "bullet"
    GRAVITY_SCALE_ELEMENT = "gravityScale"
    ANGLE_ELEMENT = "angle"
    LINEAR_DAMPING_ELEMENT = "linearDamping"
    ANGULAR_DAMPING_ELEMENT = "angularDamping"

    DEFAULT_BODY_TYPE = "StaticBody"
    DEFAULT_ACTIVE = True
    DEFAULT_AWAKE = True
    DEFAULT_ALLOW_SLEEP = True
    DEFAULT_FIXED_ROTATION = True
    DEFAULT_BULLET = False
    DEFAULT_GRAVITY_SCALE = 1.0
    DEFAULT_ANGLE = 0.0
    DEFAULT_LINEAR_DAMPING = 0.0
    DEFAULT_ANGULAR_DAMPING = 0.0

    def read_prefab_from_string(self, xml_string):
        try:
            element = EXml.fromstring(xml_string)

            body_type = self.DEFAULT_BODY_TYPE
            active = self.DEFAULT_ACTIVE
            awake = self.DEFAULT_AWAKE
            allow_sleep = self.DEFAULT_ALLOW_SLEEP
            fixed_rotation = self.DEFAULT_FIXED_ROTATION
            bullet = self.DEFAULT_BULLET
            gravity_scale = self.DEFAULT_GRAVITY_SCALE
            angle = self.DEFAULT_ANGLE
            angular_damping = self.DEFAULT_ANGULAR_DAMPING
            linear_damping = self.DEFAULT_LINEAR_DAMPING

            if element.find(self.BODY_TYPE_ELEMENT) is not None:
                body_type = element.find(self.BODY_TYPE_ELEMENT).text.strip()

            if element.find(self.ACTIVE_ELEMENT) is not None:
                active = string_to_boolean(element.find(self.ACTIVE_ELEMENT).text.strip())

            if element.find(self.AWAKE_ELEMENT) is not None:
                awake = string_to_boolean(element.find(self.AWAKE_ELEMENT).text.strip())

            if element.find(self.ALLOW_SLEEP_ELEMENT) is not None:
                allow_sleep = string_to_boolean(element.find(self.ALLOW_SLEEP_ELEMENT).text.strip())

            if element.find(self.FIXED_ROTATION_ELEMENT) is not None:
                fixed_rotation = string_to_boolean(element.find(self.FIXED_ROTATION_ELEMENT).text.strip())

            if element.find(self.BULLET_ELEMENT) is not None:
                bullet = string_to_boolean(element.find(self.BULLET_ELEMENT).text.strip())

            if element.find(self.GRAVITY_SCALE_ELEMENT) is not None:
                gravity_scale = float(element.find(self.GRAVITY_SCALE_ELEMENT).text.strip())

            if element.find(self.ANGLE_ELEMENT) is not None:
                angle = float(element.find(self.ANGLE_ELEMENT).text.strip())

            if element.find(self.ANGULAR_DAMPING_ELEMENT) is not None:
                angular_damping = float(element.find(self.ANGULAR_DAMPING_ELEMENT).text.strip())

            if element.find(self.LINEAR_DAMPING_ELEMENT) is not None:
                linear_damping = float(element.find(self.LINEAR_DAMPING_ELEMENT).text.strip())

            return BodyPrefab(body_type, active, allow_sleep, awake, bullet,
                              fixed_rotation, gravity_scale, angle, angular_damping, linear_damping)
        except Exception as e:
            raise BodyPrefabXmlReadException(xml_string, e)


class BodyPrefabXmlReadException(Exception):
    MESSAGE_TEMPLATE = "Cannot read body prefab from xml string {}. Cause: {}."

    def __init__(self, xml_string, cause):
        super().__init__(xml_string, cause)

File: core/src/remove/tanks/game/physics/body/BodyPrefabXmlReader.java
package remove.tanks.game.physics.body;

import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.utils.XmlReader;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Mateusz Długosz
 */
public final class BodyPrefabXmlReader {
    public static final String BODY_ELEMENT = "body";
    public static final String BODIES_ELEMENT = "bodies";

    private static final String BODY_TYPE_ELEMENT = "type";
    private static final String ACTIVE_ELEMENT = "active";
    private static final String AWAKE_ELEMENT = "awake";
    private static final String ALLOW_SLEEP_ELEMENT = "allowSleep";
    private static final String FIXED_ROTATION_ELEMENT = "fixedRotation";
    private static final String BULLET_ELEMENT = "bullet";
    private static final String GRAVITY_SCALE_ELEMENT = "gravityScale";
    private static final String ANGLE_ELEMENT = "angle";
    private static final String LINEAR_DAMPING_ELEMENT = "linearDamping";
    private static final String ANGULAR_DAMPING_ELEMENT = "angularDamping";

    private static final BodyDef.BodyType BODY_TYPE_DEFAULT_VALUE = BodyDef.BodyType.StaticBody;
    private static final boolean ACTIVE_DEFAULT_VALUE = true;
    private static final boolean AWAKE_DEFAULT_VALUE = true;
    private static final boolean ALLOW_SLEEP_DEFAULT_VALUE = true;
    private static final boolean FIXED_ROTATION_DEFAULT_VALUE = true;
    private static final boolean BULLET_DEFAULT_VALUE = false;
    private static final float GRAVITY_SCALE_DEFAULT_VALUE = 1f;
    private static final float ANGLE_DEFAULT_VALUE = 0f;
    private static final float LINEAR_DAMPING_DEFAULT_VALUE = 0f;
    private static final float ANGULAR_DAMPING_DEFAULT_VALUE = 0f;

    public BodyPrefab readBodyPrefab(XmlReader.Element element) {
        try {
            return new BodyPrefab(
                    readBodyType(element),
                    readActive(element),
                    readAllowSleep(element),
                    readAwake(element),
                    readBullet(element),
                    readFixedRotation(element),
                    readGravityScale(element),
                    readAngle(element),
                    readAngularDamping(element),
                    readLinearDamping(element)
            );
        } catch (Exception e) {
            throw new BodyPrefabXmlReadException(element, e);
        }
    }

    private BodyDef.BodyType readBodyType(XmlReader.Element element) {
        if (element.getChildByName(BODY_TYPE_ELEMENT) == null) {
            return BODY_TYPE_DEFAULT_VALUE;
        }
        return BodyDef.BodyType.valueOf(element.getChildByName(BODY_TYPE_ELEMENT).getText().trim());
    }

    private boolean readActive(XmlReader.Element element) {
        if (element.getChildByName(ACTIVE_ELEMENT) == null) {
            return ACTIVE_DEFAULT_VALUE;
        }
        return Boolean.valueOf(element.getChildByName(ACTIVE_ELEMENT).getText().trim());
    }

    private boolean readAllowSleep(XmlReader.Element element) {
        if (element.getChildByName(ALLOW_SLEEP_ELEMENT) == null) {
            return ALLOW_SLEEP_DEFAULT_VALUE;
        }
        return Boolean.valueOf(element.getChildByName(ALLOW_SLEEP_ELEMENT).getText().trim());
    }

    private boolean readAwake(XmlReader.Element element) {
        if (element.getChildByName(AWAKE_ELEMENT) == null) {
            return AWAKE_DEFAULT_VALUE;
        }
        return Boolean.valueOf(element.getChildByName(AWAKE_ELEMENT).getText().trim());
    }

    private boolean readBullet(XmlReader.Element element) {
        if (element.getChildByName(BULLET_ELEMENT) == null) {
            return BULLET_DEFAULT_VALUE;
        }
        return Boolean.valueOf(element.getChildByName(BULLET_ELEMENT).getText().trim());
    }

    private boolean readFixedRotation(XmlReader.Element element) {
        if (element.getChildByName(FIXED_ROTATION_ELEMENT) == null) {
            return FIXED_ROTATION_DEFAULT_VALUE;
        }
        return Boolean.valueOf(element.getChildByName(FIXED_ROTATION_ELEMENT).getText().trim());
    }

    private float readGravityScale(XmlReader.Element element) {
        if (element.getChildByName(GRAVITY_SCALE_ELEMENT) == null) {
            return GRAVITY_SCALE_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(GRAVITY_SCALE_ELEMENT).getText().trim());
    }

    private float readAngle(XmlReader.Element element) {
        if (element.getChildByName(ANGLE_ELEMENT) == null) {
            return ANGLE_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(ANGLE_ELEMENT).getText().trim());
    }

    private float readAngularDamping(XmlReader.Element element) {
        if (element.getChildByName(ANGULAR_DAMPING_ELEMENT) == null) {
            return ANGULAR_DAMPING_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(ANGULAR_DAMPING_ELEMENT).getText().trim());
    }

    private float readLinearDamping(XmlReader.Element element) {
        if (element.getChildByName(LINEAR_DAMPING_ELEMENT) == null) {
            return LINEAR_DAMPING_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(LINEAR_DAMPING_ELEMENT).getText().trim());
    }
}
